Add a POJA serverless implementation based on Apache

andriy-dmytruk opened the pull request:
The idea of this module is to provide a solution that is able to parse simple requests and give simple responses using System `STDIN` and `STDOUT` instead of networking. The common logic is present in the `http-poja-common` and there is a particular Apache implementation in `http-poja-apache`.

This is an improvement of https://github.com/micronaut-projects/micronaut-servlet/pull/748 PR that uses a more commonly-used library instead of RawHTTP. The common module can be used for implementing other clients relatively easily, though.

There is a TCK that is passing for the created module.

wwwsahoo committed:
GCN-4404: Initial commit of serverless micronaut web applications.

This is currently using rawhttp as parser. I will change it soon to use a more widely used parser.

I am able to test a sample Micronaut application with this module. The sample can be found in samples/ dir along with its own README and build script.
I can see the app working both in interactive mode as well as inetd mode.

Here's the plan before PR is submitted:
1. Use a more widely used parser.
2. Add test cases and samples for http-serverless module.
3. Fix TODOs in code.
4. Run compatibility tests for micronaut.
5. Update gcn/micronaut launcher to allow users to select the new serverless option.

andriy-dmytruk committed:
GCN-4622 Create a simple test for the serverless application

The test uses pipes to communicate with the server instead of STDIN and STDOUT and currently sends requests and receives responses as text, which can be improved.

wwwsahoo committed:
added copyright and license.

andriy-dmytruk committed:
GCN-4624 Add the Micronaut server TCK for POJA

andriy-dmytruk committed:
Add body support to the poja

- Add support for getting the body from the RawHttpRequest
- Add support for reading cookies from the request

andriy-dmytruk committed:
Improve body support

- Add a LimitingInputStream to improve body support. The stream does not allow reading after content length has been read.
- Refactor into common base classes and specific RawHttp implementation that could be replaced.

andriy-dmytruk committed:
Improvements with closing client and streams

andriy-dmytruk committed:
Store attributes in the POJA request

andriy-dmytruk committed:
Use the ServletResponseFactory for creating responses

andriy-dmytruk committed:
Change body input stream reading

- Fix the input stream reading for tests.
- Remove the transfer encoding header

andriy-dmytruk committed:
More TCK test fixes

- Make sure byte data is preserved
- Fix getting header that is missing

andriy-dmytruk committed:
Fixing and refactoring

- Make the PojaHttpRequest mutable for filters.
- Make sure that headers are standardized to upper case.
- Allow binding byte[] body.
- Move the common testing logic to the http-poja-test module.
- Use limiting input stream on the whole body to make sure too many bytes are not read.

andriy-dmytruk committed:
Move the sample

andriy-dmytruk committed:
Copy QueryStringDecoderTest from netty

andriy-dmytruk committed:
Remove TODO

andriy-dmytruk committed:
Spotless and binary compatability fixes

andriy-dmytruk committed:
Fix checkstyle

andriy-dmytruk committed:
Decouple POJA into a common and implementation module

andriy-dmytruk committed:
Begin adding Apache implementation for HTTP POJA

andriy-dmytruk committed:
Fix header parsing and body reading for Apache POJA

andriy-dmytruk committed:
Fix cookie parsing for Apache POJA

andriy-dmytruk committed:
More fixes for TCK tests

- Remove unneeded dependency.
- Fix for header reading.
- Fix for reading input stream by limiting it with content length.

andriy-dmytruk committed:
Fix the LimitingInputStream

andriy-dmytruk committed:
Fix header standardization

Fix header standardization method and refactor header and query parameters to separate classes

andriy-dmytruk committed:
Refactor, fix checkstyle and javadoc

andriy-dmytruk committed:
Minor fix for sample

andriy-dmytruk committed:
Fix TCK client configuration

andriy-dmytruk committed:
Enable CORS test

graemerocher commented on test-sample-poja/build.gradle:
```diff
@@ -0,0 +1,40 @@
+ * limitations under the License.
+ */
+plugins {
+    id("io.micronaut.build.internal.servlet.module")
```
I don't think this should be a module otherwise it will be published

graemerocher commented on test-sample-poja/src/main/java/io/micronaut/http/poja/sample/Application.java:
```diff
@@ -0,0 +1,39 @@
+        // which gets mixed with HTTP response.
+        // See GCN-4489
+        Micronaut.build(args)
+                .banner(false)
```
can we configure this in an `ApplicationContextConfigurer`

graemerocher commented on test-suite-http-server-tck-poja-apache/build.gradle:
```diff
@@ -0,0 +1,31 @@
+plugins {
+    id("io.micronaut.build.internal.servlet.module")
```
again not sure this should be a module

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServerlessApplication.java:
```diff
@@ -0,0 +1,97 @@
+    ) throws IOException {
+        ConversionService conversionService = applicationContext.getConversionService();
+        MediaTypeCodecRegistry codecRegistry = applicationContext.getBean(MediaTypeCodecRegistry.class);
+        ExecutorService ioExecutor = applicationContext.getBean(ExecutorService.class, Qualifiers.byName(TaskExecutors.BLOCKING));
```
move these lookups into fields of this class so they are not repeated

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServerlessApplication.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    private void writeResponse(ClassicHttpResponse response, OutputStream out) throws IOException {
+        SessionOutputBuffer buffer = new SessionOutputBufferImpl(8 * 1024);
```
buffer size should probably be configurable

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServerlessApplication.java:
```diff
@@ -0,0 +1,97 @@
+        try {
+            responseWriter.write(response, buffer, out);
+        } catch (HttpException e) {
+            throw new RuntimeException("Could not write response body", e);
```
throw a more specific exception like HttpServerException or something

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpRequest.java:
```diff
@@ -0,0 +1,307 @@
+ * @author Andriy Dmytruk
+ * @since 4.10.0
+ */
+public class ApacheServletHttpRequest<B> extends PojaHttpRequest<B, ClassicHttpRequest, ClassicHttpResponse> {
```
annotate `@Internal`

graemerocher replied:
can it be `final`?

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpRequest.java:
```diff
@@ -0,0 +1,307 @@
+    ) {
+        super(conversionService, codecRegistry, response);
+
+        SessionInputBufferImpl sessionInputBuffer = new SessionInputBufferImpl(8192);
```
buffer size should probably be configurable

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpRequest.java:
```diff
@@ -0,0 +1,307 @@
+        try {
+            request = parser.parse(sessionInputBuffer, inputStream);
+        } catch (HttpException | IOException e) {
+            throw new RuntimeException("Could parse HTTP request", e);
```
throw more specific exceptions

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpRequest.java:
```diff
@@ -0,0 +1,307 @@
+                bodyStream, optionalContentLength, ioExecutor
+            );
+        } catch (IOException e) {
+            throw new RuntimeException("Could not get request body", e);
```
throw more specific exceptions

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpRequest.java:
```diff
@@ -0,0 +1,307 @@
+
+    @Override
+    public <T> MutableHttpRequest<T> body(T body) {
+        return null;
```
return `this` or finish the implementation

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpRequest.java:
```diff
@@ -0,0 +1,307 @@
+
+    @Override
+    public void setConversionService(@NonNull ConversionService conversionService) {
+
```
and comment saying this is not implemented

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpRequest.java:
```diff
@@ -0,0 +1,307 @@
+    private static MultiValueHeaders createHeaders(
+        Header[] headers, ConversionService conversionService
+    ) {
+        Map<String, List<String>> map = new HashMap<>();
```
should probably be a linked hash map to maintain order

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpRequest.java:
```diff
@@ -0,0 +1,307 @@
+     * An input stream that would initially delegate to the first input stream
+     * and then to the second one. Created specifically to be used with {@link ByteBody}.
+     */
+    private static class CombinedInputStream extends InputStream {
```
should be `final`?

graemerocher commented on http-poja-apache/src/main/java/io/micronaut/http/poja/llhttp/ApacheServletHttpResponse.java:
```diff
@@ -0,0 +1,143 @@
+ * @author Andriy Dmytruk
+ * @since 4.10.0
+ */
+public class ApacheServletHttpResponse<T> extends PojaHttpResponse<T, ClassicHttpResponse> {
```
annotate with `@Internal`

andriy-dmytruk committed:
Add a test

graemerocher commented:
if you can try address the sonar code quality checks as well would be great. Thanks

andriy-dmytruk committed:
Remove sample and tck from being modules

andriy-dmytruk committed:
Create PojaHttpServerlessApplicationContextConfigurer

andriy-dmytruk committed:
Create more specific exceptions

Handle the case of a bad request where it could not be parsed. The server should not break, but return a 400 response. Also implement other review comments.

andriy-dmytruk committed:
Add buffer size configuration options

andriy-dmytruk committed:
Fix sonar cloud warnings

andriy-dmytruk committed:
Minor style fix

andriy-dmytruk committed:
More sonar cloud fixes

andriy-dmytruk committed:
Fix TCK test with native-image

andriy-dmytruk committed:
Attempt to fix GraalVM 17 native image

andriy-dmytruk committed:
Revert and fix GraalVM TCK by adding required configurations
